clash-cores/src/Clash/Cores/Crc/Internal.hs

```diff
@@ -0,0 +1,704 @@
+    g = postProcess $ bv2v . (\p -> runCrc 0 p) <$> onehots dataWidth
+
+-- | Vertically partition a matrix
+partitionMat
+  :: KnownNat q
+  => SNat p
+  -> Vec n (BitVector (p * q))
+  -> Vec n (Vec p (BitVector q))
+partitionMat SNat mat = fmap unpack mat
+
+-- | Flatten vertically partitioned matrix
+unPartitionMat
+  :: KnownNat q
+  => KnownNat p
+  => Vec n (Vec p (BitVector q))
+  -> Vec n (BitVector (p * q))
+unPartitionMat mat = fmap pack mat
+
+-- | Helper type that contains the flattened @F@ and @G@ matrices for each lane
+data CrcLaneParams (crcWidth :: Nat) (dataWidth :: Nat) (nLanes :: Nat) where
+  -- | The final lane
+  CrcLane1
+    :: Vec crcWidth (BitVector (crcWidth + dataWidth))
+    -- ^ Flattened @F|G@ matrix for this lane
+    -> CrcLaneParams crcWidth dataWidth 1
+  -- | Not the final lane
+  CrcLaneN
+    :: 2 <= lane
+    => SNat lane
+    -- ^ Which lane this is
+    -> Vec crcWidth (BitVector (crcWidth + lane * dataWidth))
+    -- ^ Flattened @F|G@ matrix for this lane
+    -> CrcLaneParams crcWidth dataWidth (lane - 1)
+    -> CrcLaneParams crcWidth dataWidth lane
+
+deriving instance (KnownNat crcWidth, KnownNat dataWidth, KnownNat nLanes)
+  => Lift (CrcLaneParams crcWidth dataWidth nLanes)
+
+deriving instance (KnownNat crcWidth, KnownNat dataWidth, KnownNat nLanes)
+  => Show (CrcLaneParams crcWidth dataWidth nLanes)
+
+-- | No domain in 'CrcLaneParams'
+type instance TryDomain t (CrcLaneParams crcWidth dataWidth nLanes) = 'NotFound
+
+-- | Contains all necessary parameters for the hardware CRC implementation
+data CrcHardwareParams (crcWidth :: Nat) (dataWidth :: Nat) (nLanes :: Nat)
+  = CrcHardwareParams
+      { _crcDataWidthHW :: SNat dataWidth
+      , _crcNlanes :: SNat nLanes
+      , _crcBaseParams :: CrcParams crcWidth
+      , _crcLaneParams :: CrcLaneParams crcWidth dataWidth nLanes
+      , _crcResidues :: Vec nLanes (BitVector crcWidth)
+      }
+  deriving (Show, Lift)
+
+-- | No domain in 'CrcHardwareParams'
+type instance TryDomain t (CrcHardwareParams crcWidth dataWidth nLanes) = 'NotFound
+
+-- | This class is used to indicate a CRC has a derived hardware implementation
+--
+-- @nLanes@ indicates the maximum number of @dataWidth@ words you want to
+-- process in a single cycle. For example the stream could be byte-oriented,
+-- but processing is done @n@ bytes at a time.
+--
+-- Use the 'Clash.Cores.Crc.deriveHardwareCrc' to create an instance.
+-- No instances should be implemented by hand, because a proper instance
+-- requires compile-time evaluation.
+class
+  ( KnownCrc crc, KnownNat dataWidth, 1 <= dataWidth, KnownNat nLanes, 1 <= nLanes)
+  => HardwareCrc (crc :: Type) (dataWidth :: Nat) (nLanes :: Nat) where
+  crcHardwareParams
+    :: crc
+    -- ^ Which CRC
+    -> Proxy dataWidth
+    -- ^ What word size in bits the hardware implementation can handle
+    -> Proxy nLanes
+    -- ^ The number of lanes
+    -> CrcHardwareParams (CrcWidth crc) dataWidth nLanes
```

Comment:
I like `SNat (n :: Nat)` better than `Proxy (n :: Nat)`. It's there, why not use it. This seems to compile fine:

```diff
--- a/clash-cores/src/Clash/Cores/Crc/Internal.hs
+++ b/clash-cores/src/Clash/Cores/Crc/Internal.hs
@@ -394,9 +394,9 @@ class
   crcHardwareParams
     :: crc
     -- ^ Which CRC
-    -> Proxy dataWidth
+    -> SNat dataWidth
     -- ^ What word size in bits the hardware implementation can handle
-    -> Proxy nLanes
+    -> SNat nLanes
     -- ^ The number of lanes
     -> CrcHardwareParams (CrcWidth crc) dataWidth nLanes
 
@@ -623,7 +623,7 @@ crcEngine
   --
   -> Signal dom (BitVector (CrcWidth crc))
   -- ^ The resulting CRC. There is a delay of a single cycle from input to output.
-crcEngine crc = crcEngineFromParams $ crcHardwareParams crc Proxy Proxy
+crcEngine crc = crcEngineFromParams $ crcHardwareParams crc SNat SNat
 
 -- | Like 'crcEngine' but uses 'CrcHardwareParams' instead of a 'HardwareCrc' constraint
 crcEngineFromParams
@@ -675,7 +675,7 @@ crcValidator
   -- ^ The input data
   -> Signal dom Bool
   -- ^ Whether the CRC is valid. There is a delay of a single cycle from input to output.
-crcValidator crc = crcValidatorFromParams $ crcHardwareParams crc Proxy Proxy
+crcValidator crc = crcValidatorFromParams $ crcHardwareParams crc SNat SNat
 
 -- | Like 'crcValidator' but uses 'CrcHardwareParams' instead of a 'HardwareCrc' constraint
 crcValidatorFromParams
```

Reply:
In addition, you could drop the `KnownNat dataWidth` and `KnownNat nLanes`.

Reply:
Heck, if you try hard enough, you can even get rid of the `KnownNat crcWidth`! But is there a way to use `{..}` and pattern matching in one go? It would eliminate the function `go` in this suggestion:

```diff
--- a/clash-cores/src/Clash/Cores/Crc/Internal.hs
+++ b/clash-cores/src/Clash/Cores/Crc/Internal.hs
@@ -657,15 +657,17 @@ crcEngineFromParams
             (dataWidth :: Nat)
             (nLanes :: Nat)
    . HiddenClockResetEnable dom
-  => KnownNat crcWidth
-  => KnownNat dataWidth
-  => KnownNat nLanes
   => CrcHardwareParams crcWidth dataWidth nLanes
   -> Signal dom Bool
   -> Signal dom (Maybe (Index nLanes, Vec nLanes (BitVector dataWidth)))
   -> Signal dom (BitVector crcWidth)
-crcEngineFromParams hwParams@(CrcHardwareParams _ _ CrcParams{..} _ _) resetCrc inDat = crcOut
-    where
+crcEngineFromParams
+  hwParams@(CrcHardwareParams SNat SNat _crcBaseParams@(CrcParams SNat _ _ _ _ _) _ _)
+  resetCrc inDat
+  = go _crcBaseParams
+   where
+    go CrcParams{..} = crcOut
+     where
       steps = smap (\sn _ -> laneStep hwParams sn) (repeat @nLanes ())
       (validLanesX, datX) = unbundle $ fromJustX <$> inDat
       step = fmap (steps !!) validLanesX
@@ -709,16 +711,18 @@ crcValidatorFromParams
             (dataWidth :: Nat)
             (nLanes :: Nat)
    . HiddenClockResetEnable dom
-  => KnownNat crcWidth
-  => KnownNat dataWidth
-  => KnownNat nLanes
   => 1 <= nLanes
   => CrcHardwareParams crcWidth dataWidth nLanes
   -> Signal dom Bool
   -> Signal dom (Maybe (Index nLanes, Vec nLanes (BitVector dataWidth)))
   -> Signal dom Bool
-crcValidatorFromParams hwParams@(CrcHardwareParams _ _ CrcParams{..} _ residues) resetCrc inDat = match
-    where
+crcValidatorFromParams
+  hwParams@(CrcHardwareParams SNat SNat _crcBaseParams@(CrcParams SNat _ _ _ _ _) _ residues)
+  resetCrc inDat
+  = go _crcBaseParams
+   where
+    go CrcParams{..} = match
+     where
       step = laneStep hwParams (subSNatLe (SNat @nLanes) d1)
       (validLanesX, datX) = unbundle $ fromJustX <$> inDat
```

This is all a bit splitting hairs, though, these functions are internal... but I do like to split hairs, so thanks :-)